### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by ayrat555**
Improvements:

- [x] Format code

- [x] Fix circle ci config
  - fail ci if dialyzer fails
  - more caching

- [ ] Make dialyzer happy

### CODE REVIEW

**Comment by masonforest** [comment]
Looks good to me 👍 


Looks like Circle is still failing though?

**Comment by ayrat555** [comment]
@masonforest yep, I didn't fix dialyzer yet

File: lib/block/header.ex [hunk 1]
```diff
@@ -84,9 +104,9 @@ defmodule Block.Header do
       h.receipts_root,
       h.logs_bloom,
       h.difficulty,
-      (if h.number == 0, do: <<>>, else: h.number),
```
Comment (hayesgm) [inline lib/block/header.ex]: I don't understand this syntax change.

File: lib/block/header.ex [hunk 2]
```diff
@@ -3,56 +3,76 @@ defmodule Block.Header do
...
+          # TODO: 64-bit hash?
+          nonce: <<_::64>> | nil
+        }
 
   # The start of the Homestead block, as defined in Eq.(13) of the Yellow Paper (N_H)
   @homestead_block 1_150_000
 
-  @initial_difficulty 131_072 # d_0 from Eq.(40)
-  @minimum_difficulty @initial_difficulty # Mimics d_0 in Eq.(39), but variable on different chains
+  # d_0 from Eq.(40)
```
Comment (hayesgm) [inline lib/block/header.ex]: These comments (I think) made more sense inline

File: lib/block/header.ex [hunk 3]
```diff
@@ -3,56 +3,76 @@ defmodule Block.Header do
...
-    gas_used: 0,                     # Hg
-    timestamp: nil,                  # Hs
-    extra_data: <<>>,                # Hx
-    mix_hash: nil,                   # Hm
-    nonce: nil,                      # Hn
-  ]
+  @empty_trie MerklePatriciaTree.Trie.empty_trie_root_hash()
+  @empty_keccak [] |> ExRLP.encode() |> :keccakf1600.sha3_256()
+
+            # Hp P(BH)Hr
```
Comment (hayesgm) [inline lib/block/header.ex]: Is it possible to control the settings on the formatter?

File: .circleci/config.yml [hunk 4]
```diff
@@ -7,14 +7,42 @@ jobs:
       - checkout
       - run: git submodule sync --recursive
       - run: git submodule update --recursive --init
+
       - restore_cache:
-          key: _build
+          keys:
+            - v1-dependency-cache-{{ arch }}-{{ checksum "mix.lock" }}
+            - v1-dependency-cache-{{ arch }}
```
Comment (hayesgm) [inline .circleci/config.yml]: Can you explain exactly what this accomplishes over the previous?

File: lib/block/header.ex [hunk 5]
```diff
@@ -251,17 +272,64 @@ defmodule Block.Header do
...
     parent_gas_limit = if parent_header, do: parent_header.gas_limit, else: nil
 
-    errors = [] ++
-      (if header.difficulty == get_difficulty(header, parent_header, initial_difficulty, minimum_difficulty, difficulty_bound_divisor, homestead_block), do: [], else: [:invalid_difficulty]) ++ # Eq.(51)
-      (if header.gas_used <= header.gas_limit, do: [], else: [:exceeded_gas_limit]) ++ # Eq.(52)
-      (if is_gas_limit_valid?(header.gas_limit, parent_gas_limit, gas_limit_bound_divisor, min_gas_limit), do: [], else: [:invalid_gas_limit]) ++ # Eq.(53), Eq.(54) and Eq.(55)
-      (if is_nil(parent_header) or header.timestamp > parent_header.timestamp, do: [], else: [:child_timestamp_invalid]) ++ # Eq.(56)
-      (if header.number == 0 or header.number == parent_header.number + 1, do: [], else: [:child_number_invalid]) ++ # Eq.(57)
-      (if byte_size(header.extra_data) <= @max_extra_data_bytes, do: [], else: [:extra_data_too_large])
+    # Eq.(51)
```
Comment (hayesgm) [inline lib/block/header.ex]: Again, these comments are broken.

Comment (hayesgm) [reply lib/block/header.ex]: And (due to the formatting) this function would need to be refactored.

File: .circleci/config.yml (hunk 4 above)
Comment (ayrat555) [reply .circleci/config.yml]: @hayesgm I added cache fallbacks. so if ```v1-dependency-cache-{{ arch }}-{{ checksum "mix.lock" }}``` cache does not exist, it fallbacks to ```v1-dependency-cache-{{ arch }}``` and etc.

Recently I encountered a problem with elixir and circle ci on another project. Circle CI sometimes fails to find cache even if ```mix.lock``` (and its hash respectively) did not change. Caching is very critical for dialyzer check speed. Adding ```arch``` template key solves the problem.

https://discuss.circleci.com/t/use-the-arch-cache-template-key-if-you-rely-on-cached-compiled-binary-dependencies/16129

File: lib/block/header.ex (hunk 2 above)
Comment (ayrat555) [reply lib/block/header.ex]: @hayesgm yes, I agree inline comments look better and they are more readable. But elixir formatter does not think so :). It does not have many configuration parameters. I think it was made that way to get rid of code style debates. 

It even formats inline comments not equivalently. I went over all changed files checking and fixing comments but unfortunately missed some broken comments. I will fix them.

File: lib/block/header.ex (hunk 1 above)
Comment (ayrat555) [reply lib/block/header.ex]: @hayesgm ```if``` is a method that takes boolean expression, and keyword as parameters. so formatter thinks it should look better like ```if(h.number == 0, do: <<>>, else: h.number)```

File: lib/block/header.ex (hunk 3 above)
Comment (ayrat555) [reply lib/block/header.ex]: @hayesgm I did not find this setting in [official docs](https://hexdocs.pm/mix/master/Mix.Tasks.Format.html)

File: lib/block/header.ex (hunk 1 above)
Comment (hayesgm) [reply lib/block/header.ex]: oh cool, though it must be a macro, not a method. In either case, this appears less clear than the previous code.

**Comment by ayrat555** [comment]
@hayesgm I returned ```header.ex``` to original state for now. I will refactore it soon in next PRs. I think ```is_valid?``` method should be refactored in more functional style:
```
  @spec is_valid?(...) :: :ok | {:error, atom()}
  def is_valid?(...) do
    with :ok <- check_header_difficulty(...)
         :ok <- check_gas(...)
    ...
  end
```

btw I think it's better not to use `ifs` to make code less imperative and more concise.
```

I can split this PR's changes into multuple PRs if you want but there are no drastic changes. The only thing that left to fix is typespecs.

**Pull request merged**